File: content/snippets/go/interview_lcci/17/07_baby_names.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

var parent map[string]string

func find(x string) string {
	if _, ok := parent[x]; !ok {
		parent[x] = x
	}
	if parent[x] != x {
		parent[x] = find(parent[x])
	}
	return parent[x]
}

func unite(a, b string) {
	pa, pb := find(a), find(b)
	if pa != pb {
		if pa < pb {
			parent[pb] = pa
		} else {
			parent[pa] = pb
		}
	}
}

func trulyMostPopular(names []string, synonyms []string) []string {
	parent = make(map[string]string)
	freq := make(map[string]int)
	for _, s := range names {
		paren := strings.Index(s, "(")
		name := s[:paren]
		count, _ := strconv.Atoi(s[paren+1 : len(s)-1])
		freq[name] = count
	}
	for _, syn := range synonyms {
		s := syn[1 : len(syn)-1]
		comma := strings.Index(s, ",")
		unite(s[:comma], s[comma+1:])
	}
	result := make(map[string]int)
	for name, cnt := range freq {
		result[find(name)] += cnt
	}
	ans := make([]string, 0, len(result))
	for name, cnt := range result {
		ans = append(ans, name+"("+strconv.Itoa(cnt)+")")
	}
	sort.Strings(ans)
	return ans
}

func main() {
	names := []string{"John(15)", "Jon(12)", "Chris(13)", "Kris(4)", "Christopher(19)"}
	synonyms := []string{"(Jon,John)", "(John,Johnny)", "(Chris,Kris)", "(Chris,Christopher)"}
	fmt.Println(strings.Join(trulyMostPopular(names, synonyms), " ")) // Chris(36) John(27)
}

File: content/snippets/python/interview_lcci/17/07_baby_names.py
from typing import List
from collections import defaultdict


class Solution:
    def trulyMostPopular(self, names: List[str], synonyms: List[str]) -> List[str]:
        parent: dict = {}

        def find(x: str) -> str:
            if x not in parent:
                parent[x] = x
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def unite(a: str, b: str) -> None:
            pa, pb = find(a), find(b)
            if pa != pb:
                if pa < pb:
                    parent[pb] = pa
                else:
                    parent[pa] = pb

        freq: dict = {}
        for s in names:
            paren = s.index('(')
            name = s[:paren]
            count = int(s[paren + 1:-1])
            freq[name] = count

        for syn in synonyms:
            s = syn[1:-1]
            comma = s.index(',')
            unite(s[:comma], s[comma + 1:])

        result: dict = defaultdict(int)
        for name, cnt in freq.items():
            result[find(name)] += cnt

        return sorted(f"{name}({cnt})" for name, cnt in result.items())


def main():
    sol = Solution()
    names = ["John(15)", "Jon(12)", "Chris(13)", "Kris(4)", "Christopher(19)"]
    synonyms = ["(Jon,John)", "(John,Johnny)", "(Chris,Kris)", "(Chris,Christopher)"]
    print(' '.join(sol.trulyMostPopular(names, synonyms)))  # Chris(36) John(27)


if __name__ == "__main__":
    main()
